src/main/java/io/vertx/core/streams/WriteStream.java

```diff
@@ -68,6 +70,11 @@ default void end(T t) {
     end();
   }
 
+  default void end(Handler<AsyncResult<Void>> handler) {
+    end();
+    handler.handle(Future.succeededFuture());
+  }
+
   /**
```

Comment:
Would you not need to handle the case where the end() call fails and therefore you need to pass a Future.failedFuture() to the handler?

Reply:
Since it's default implementation I tried to preserve original behaviour of `end()`. If it throws - caller get the exception. Same here in `end(Handler)`. Do you think I should do something like:
```java
  default void end(Handler<AsyncResult<Void>> handler) {
    Future<Void> fut = Future.<Void>future();
    try {
      end();
      fut.complete();
    } catch (Exception e) {
      fut.fail(e);  
    }
    handler.handle(fut);
  }

```

Reply:
yes

Reply:
we should not provide a default implementation because `end()` is asynchronous, so the default implementation cannot assume that the callback can be called after the end() method call

Reply:
in addition, the write stream may be full and the actual end completion might be delayed until the stream is drained.  Check the `AsyncFile` implementation that delays the actual close until the pending buffers are written.

Reply:
Ok, I get: `end()` is asynchronous, it just didn't signal the actual completion. I think this should be reflected in the javadoc of this method (much like in one's for `write()`).

But original problem still stays: some classes (like `StreamingBodyCodec` in vertx-web-client) need to get notification when `WriteStream` end happens. And there is no way to achieve this with the current interface. That's why I'm proposing to change the interface: add `WriteStream.end(Handler)`. To me this looks like the most consistent way of solving the problem. But I admit that I'm not very familiar with Vert.x internals. Is there better way?
@vietj 
@dgoetsch

Reply:
can you provide a reproducer for the WebClient issue ?

Reply:
@mkulak that sounds like precisely the use case for an `endHandler`.

Reply:
before speculating on a fix, it would be good to have a reproducer for the `WebClient` issue

Reply:
[Real world test](https://gist.github.com/mkulak/764536e26ade3f26e1daf38460b413d3) (in Kotlin)
Not sure if it will fail 100% of time on every machine. 

[Simulated test](https://gist.github.com/mkulak/764536e26ade3f26e1daf38460b413d3) (in Java)
It's synthetic but always fails. For this test I make my own implementation of `AsyncFile` with `end()` that calls `close()` and `close()` that calls `close((Handler) null)` (much like real `AsyncFileImpl`). But `close(Handler)` actually delays it's execution for 3 seconds.